Reorganize and fill out documentation

rustprooflabs opened the pull request:
These initial commits start using `mdbook` to create the output and establish the initial outline of structure.  The existing architecture and designing for trust documents were brought in, no other significant content added yet.

rustprooflabs commented:
To preview the mdbook version of these docs:

```bash
cd plrust/doc
mdbook serve --open
```

The following screenshot shows the current table of contents defined by `doc/src/SUMMARY.md`.  The parent PL/Rust (e.g. home page) has the intro from the current readme.  The structure is trivial to re-arrange, suggestions welcome.

The Design and Architecture pieces were existing MD files brought over as-is. The "Install PL/Rust for Development" section has the main steps to get started with Postgres installed via apt with https://github.com/tcdi/plrust/pull/193 resolved.  I started adding basic usage examples to section 1.

![image](https://user-images.githubusercontent.com/3085224/215933104-b51885ba-eb25-43fc-81b0-647f1cd5fff6.png)

rustprooflabs commented:
Rearranged the ToC a bit, made home page title "What is PL/Rust?"  Not sure I like a question there, but when it was just "PL/Rust" it didn't seem obvious that it was a page, instead looking like a title.

The Development install instructions were adjusted a bit, I verified they works using my typical deployment for dev Pg instances.  Probably need to clarify that this is "Postgres Development" focused, not PL/Rust / pgx development focused  instructions.
The "Getting Started with PL/Rust" section has a collection of basic queries.  This includes basic details about `STRICT` and shows an example using `match` with a function's input.  What is in there is already pushing the bounds of a "getting started" section, so it won't be long until that structure starts splitting out into more targeted sections.

Both of these pages need more explanation throughout, I think everything shown is accurate and functional.

rustprooflabs commented:
Updated ToC, can see the [current version here](https://github.com/rustprooflabs/plrust/blob/a880bac26138fcefdc730a268a361966e4e6f8a3/doc/src/SUMMARY.md).  The new pages are empty or nearly empty, will be populated in near future.

rustprooflabs commented:
@eeeebbbbrrrr 
[Current ToC](https://github.com/rustprooflabs/plrust/blob/ac8b1c0aaf32a82d7c9878e4778c29636e048f21/doc/src/SUMMARY.md) on GitHub.  The rendered ToC is shown in the screenshot below.  That ToC link will let you navigate on GH, though best served via `mdbook`.

I restructured ToC slightly to put SPI, Triggers and Additional Features under Built-in functions section.  After looking closer at the source that seems to appropriate.  I added basic content illustrating current direction to Data Types, Built-in functions, and SPI sections.

On the data types page I tried splitting the table from pgx's docs into two tables, one with plain Rust types and the other with pgx types.  That feels like a helpful distinction but could just be a distraction.  Under Specifics I added one example for documenting the `Numeric` / `AnyNumeric` types.  If the basic structure looks decent I can move forward adding the rest of the types.

The Built-in functions page outlines the way I thought to group functions from the source files.  Is this the direction you want this page to go?

----

![image](https://user-images.githubusercontent.com/3085224/216780309-6a5792c9-a9e7-48be-abbc-584d54ac17de.png)

workingjubilee commented on doc/src/data-types.md:
```diff
@@ -0,0 +1,80 @@
+## Specifics
+
+
+### Numeric support
+
+The `NUMERIC` PostgreSQL data type can map to either
+[`pgx::AnyNumeric`](https://docs.rs/pgx/latest/pgx/datum/numeric/struct.AnyNumeric.html)
+or
+[`pgx::Numeric<P, S>`](https://docs.rs/pgx/latest/pgx/datum/numeric/struct.Numeric.html)
+in PL/Rust.
+
+The `pgx::AnyNumeric` type is the PostgreSQL `NUMERIC` type with default
+precision and scale values.  Generally, this is the type you’ll want to use as function arguments when working with numeric data.
+
+The `pgx::Numeric<P, S>` type is a wrapper around the PostgreSQL
+`NUMERIC(P, S)` type. Its Precision and Scale values are known at compile-time to assist with scale conversions and general type safety.
```
Something we should ask ourselves re: PL/Rust, and feel free to weigh in with your opinion:

We have `Numeric<P, S>` because it will prove more useful in pgx when working with table access, etc., and in general cases where `Numeric` of a known scaling quantity is used. However, in PL/Rust, the function in/out, and thus what programmers will often be working with, will always be implicitly `AnyNumeric` (even if `Numeric` is requested) in the Postgres ABI. So maybe in PL/Rust, we should instead go with

```rust
type Numeric = pgx::AnyNumeric;
type PreciselyScaledNumeric<P,S> = pgx::Numeric<P, S>;
```
or would that be more confusing?

eeeebbbbrrrr replied:
Hmm.  Do we want to deviate too much from pgx?

Any changes like this we make in plrust means we gotta keep it forever.  Dunno if that's good or bad, but like, if your thought was along the lines of "lets give the plrust type name the same name as SQL", then I dunno if that's much of an argument?

rustprooflabs replied:
I like `Numeric<P, S>`, it maps nicely to the SQL `NUMERIC(P, S)`.  This feels natural to me.

workingjubilee replied:
Just thought I'd check things against other people's intuitions to make sure. Then we're fine as is~

workingjubilee commented on doc/src/SUMMARY.md:
```diff
@@ -0,0 +1,23 @@
+# Summary
+
+[What is PL/Rust?](./plrust.md)
```
To respond to your concern about whether this should be a question: I like the implied "question and answer" format here!

workingjubilee reviewed (commented):
This is looking excellent overall!

rustprooflabs commented:
[Current ToC](https://github.com/rustprooflabs/plrust/blob/1d98baf28b35fdc80215a6314af55f0ae117de9b/doc/src/SUMMARY.md).

## Functions

I started filling out Datum descriptions on the out the [Built in Functions](https://github.com/rustprooflabs/plrust/blob/1d98baf28b35fdc80215a6314af55f0ae117de9b/doc/src/built-in-functions.md) page.  If that basic structure looks good I'll proceed, and do my best to add missing descriptions to pgx where they don't currently exist.  I'm also not trying to fully replicate what's on docs.rs, e.g. `FromDatum` and `IntoDatum` have a shared description.  My thought is these docs aren't where those precise details are needed.

The "Additional Features" section under functions is renamed to [Logging to PostgreSQL from PL/Rust](https://github.com/rustprooflabs/plrust/blob/1d98baf28b35fdc80215a6314af55f0ae117de9b/doc/src/logging.md).  Added a couple functional examples, the rest should come together easily enough.

The [SPI example](https://github.com/rustprooflabs/plrust/blob/1d98baf28b35fdc80215a6314af55f0ae117de9b/doc/src/spi.md) is closer to a functional example, see #200.  The page also references that `RETURNS TABLE` isn't supported per #36.

Are there other function related sections that haven't been mentioned in the current structure?  I am thinking the Datum and fcinfo sections from the [Built in Functions](https://github.com/rustprooflabs/plrust/blob/1d98baf28b35fdc80215a6314af55f0ae117de9b/doc/src/built-in-functions.md) page will go into their own sub-sections soon.

## Empty sections

The remaining sections with no helpful information yet are Production install, Trusted/Untrusted, and Triggers.  I'll continue working on getting at least a shell in each of these sections.

rustprooflabs commented:
[Current ToC](https://github.com/tcdi/plrust/blob/2f7e43e6609c94f795cdcf337c0124eefb99ef08/doc/src/SUMMARY.md)

Added notes [from deploying](https://github.com/tcdi/plrust/blob/2f7e43e6609c94f795cdcf337c0124eefb99ef08/doc/src/install-plrust-dev.md#update-development-plrust-installation) fix in #202 as an example.

Updated [SPI section](https://github.com/tcdi/plrust/blob/2f7e43e6609c94f795cdcf337c0124eefb99ef08/doc/src/spi.md)   to have functional example.

rustprooflabs commented:
[Current ToC](https://github.com/tcdi/plrust/blob/c80ca8591d170d478f759a8128bffdfe5ef076af/doc/src/SUMMARY.md) on GitHub.  The screenshot at the end shows the rendered mdbook ToC.

These latest changes included basic notes showing the difference between untrusted and trusted installation on the [production install](https://github.com/tcdi/plrust/blob/c80ca8591d170d478f759a8128bffdfe5ef076af/doc/src/install-plrust-prod.md) page.  I think the basic steps are accurate, however I have been unable to create a function using this method.  I will open a ticket for that next.

There are some rough details on the [Triggers](https://github.com/tcdi/plrust/blob/c80ca8591d170d478f759a8128bffdfe5ef076af/doc/src/triggers.md) page.  I plan to wait until #210 is resolved before working forward on that section.

Examples are being created in a `plrust` schema instead of the default `public` schema.  I like the ability to run examples and cleanly leave them in the past with `drop schema` later.

This pulls in more content to the book format, I am removing from their source files as I go.  The content I grabbed this time was the configuration details from the root README.md and the lint details from the readme.md and the entire LINT.md file.  I split the [config details](https://github.com/tcdi/plrust/blob/c80ca8591d170d478f759a8128bffdfe5ef076af/doc/src/config.md) into 3 sections for [Postgres config](https://github.com/tcdi/plrust/blob/c80ca8591d170d478f759a8128bffdfe5ef076af/doc/src/config-pg.md), [lints](https://github.com/tcdi/plrust/blob/c80ca8591d170d478f759a8128bffdfe5ef076af/doc/src/config-lints.md), and [env vars](https://github.com/tcdi/plrust/blob/c80ca8591d170d478f759a8128bffdfe5ef076af/doc/src/config-env-var.md).

![image](https://user-images.githubusercontent.com/3085224/219876117-e53b37b5-4b91-49d9-ac09-675523420ade.png)

rustprooflabs commented:
[Current ToC](https://github.com/tcdi/plrust/blob/732cc66c2b53a152198d908e6ed2ff1c770b474a/doc/src/SUMMARY.md)

Installation instructions consolidated into one doc for trusted & untrusted.  Added notes about rustc and rustup versions.  The trusted install steps work from start to finish on Ubuntu 22.04 with PostgreSQL 15 installed via `apt`.  Moved steps to update `plrust` to a separate page, added shell for cross compilation.

![image](https://user-images.githubusercontent.com/3085224/219973023-dca54a4f-22e7-4223-8456-d0cbf8026734.png)
